trie/leavesRetriever/leavesRetriever.go

```diff
@@ -0,0 +1,153 @@
+package leavesRetriever
+
+import (
+	"context"
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-core-go/hashing"
+	"github.com/multiversx/mx-chain-core-go/marshal"
+	"github.com/multiversx/mx-chain-go/common"
+	"github.com/multiversx/mx-chain-go/trie/leavesRetriever/dfsTrieIterator"
+)
+
+type leavesRetriever struct {
+	iterators      map[string]common.DfsIterator
+	lruIteratorIDs [][]byte
+	db             common.TrieStorageInteractor
+	marshaller     marshal.Marshalizer
+	hasher         hashing.Hasher
+	size           uint64
+	maxSize        uint64
+	mutex          sync.RWMutex
+}
+
+// NewLeavesRetriever creates a new leaves retriever
+func NewLeavesRetriever(db common.TrieStorageInteractor, marshaller marshal.Marshalizer, hasher hashing.Hasher, maxSize uint64) (*leavesRetriever, error) {
+	if check.IfNil(db) {
+		return nil, ErrNilDB
+	}
+	if check.IfNil(marshaller) {
+		return nil, ErrNilMarshaller
+
+	}
+	if check.IfNil(hasher) {
+		return nil, ErrNilHasher
+	}
+
+	return &leavesRetriever{
+		iterators:      make(map[string]common.DfsIterator),
+		lruIteratorIDs: make([][]byte, 0),
+		db:             db,
+		marshaller:     marshaller,
+		hasher:         hasher,
+		size:           0,
+		maxSize:        maxSize,
+	}, nil
+}
+
+// GetLeaves retrieves the leaves from the trie. If there is a saved checkpoint for the iterator id, it will continue to iterate from the checkpoint.
+func (lr *leavesRetriever) GetLeaves(numLeaves int, rootHash []byte, iteratorID []byte, ctx context.Context) (map[string]string, []byte, error) {
```

Comment:
should we set some max limit to `numLeaves` here?

Reply:
I do not think we need that. We give a `ctx` as argument, and it can retrieve data either until the context expired or until the num leaves have been retrieved.

Reply:
Added a TODO as a reminder to add a maxSize for each iterator, not only for the leavesRetriever